circuit/src/gadgets/main.rs

```diff
@@ -0,0 +1,950 @@
+	) -> Result<Self::Output, Error> {
+		// We should satisfy the equation below
+		// 1 - x * x_inv = 1
+		// x * x_inv = 0
+		// x * x_inv + res - 1 = 0
+
+		// Witness layout:
+		// | A   | B     | C   | D   | E  |
+		// | --- | ----- | --- | --- | ---|
+		// | x   | x_inv | res |     |    |
+
+		let (zero, x_inv, res) = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let x_inv = self.x.clone().value().map(|v| v.invert().unwrap_or(F::zero()));
+				let res = Value::known(F::one()) - self.x.clone().value().cloned() * x_inv.clone();
+
+				let mut ctx = RegionCtx::new(region, 0);
+				let zero = ctx.assign_advice(common.advice[0], Value::known(F::zero()))?;
+				let x_inv = ctx.assign_advice(common.advice[1], x_inv)?;
+				let res = ctx.assign_advice(common.advice[2], res)?;
+				Ok((zero, x_inv, res))
+			},
+		)?;
+
+		// [a, b, c, d, e]
+		let advices = [self.x.clone(), x_inv, res.clone(), zero.clone(), zero];
+		// [s_a, s_b, s_c, s_d, s_e, s_mul_ab, s_mul_cd, s_constant]
+		let fixed =
+			[F::zero(), F::zero(), F::one(), F::zero(), F::zero(), F::one(), F::zero(), -F::one()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
+
+		Ok(res)
+	}
+}
+
+/// Chip for select operation
+pub struct SelectChipset<F: FieldExt> {
+	bit: AssignedCell<F, F>,
+	x: AssignedCell<F, F>,
+	y: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt> SelectChipset<F> {
+	/// Create new SelectChipset
+	pub fn new(bit: AssignedCell<F, F>, x: AssignedCell<F, F>, y: AssignedCell<F, F>) -> Self {
+		Self { bit, x, y }
+	}
+}
+
+impl<F: FieldExt> Chipset<F> for SelectChipset<F> {
+	type Config = MainConfig;
+	type Output = AssignedCell<F, F>;
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, config: &Self::Config, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		// We should satisfy the equation below with bit asserted condition flag
+		// c (x - y) + y - res = 0
+		// cond * x - cond * y + y - res = 0
+
+		// Witness layout:
+		// | A   | B   | C | D   | E  |
+		// | --- | --- | - | --- | ---|
+		// | c   | x   | c | y   | res|
+
+		let res = layouter.assign_region(
+			|| "assign values",
+			|region| {
+				let res = self.x.value().zip(self.y.value()).zip(self.bit.value()).map(
+					|((x, y), bit)| {
+						if *bit == F::one() {
+							*x
+						} else {
+							*y
+						}
+					},
+				);
+				let mut ctx = RegionCtx::new(region, 0);
+				let res = ctx.assign_advice(common.advice[0], res)?;
+				Ok(res)
+			},
+		)?;
+
+		// [a, b, c, d, e]
+		let advices = [self.bit.clone(), self.x, self.bit.clone(), self.y, res.clone()];
+		// [s_a, s_b, s_c, s_d, s_e, s_mul_ab, s_mul_cd, s_constant]
+		let fixed =
+			[F::zero(), F::zero(), F::zero(), F::one(), -F::one(), F::one(), -F::one(), F::zero()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
```

Comment:
We need to constrain `self.bit` to be a boolean

Reply:
Done in https://github.com/eigen-trust/protocol/pull/132/commits/e9863edb37807f36e4db6ec5440bb31f4760e728